File: Assets/Script/Loader/ResourcesLoader.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace Hukiry {

    public class ResourcesLoader {

        public static void LoadAsync(string assetName, OnAssetLoadedCallback<UnityEngine.Object> OnLoadedCallback) {
            AssetBundleLoader.Instance.StartCoroutine(LoadAssetAsync(assetName, OnLoadedCallback));
        }

        //加载资源
        static IEnumerator LoadAssetAsync(string assetName, OnAssetLoadedCallback<UnityEngine.Object> onLoadedCallback) {
            ResourceRequest rr = Resources.LoadAsync(assetName);
            while (!rr.isDone) {
                yield return new WaitForEndOfFrame();
            }
            onLoadedCallback(assetName, rr.asset);
            yield return null;
        }

        public static Object LoadAsset(string assetName)
        {
           return Resources.Load(assetName);
        }

        public static void Unload(UnityEngine.Object obj)
        {
            Resources.UnloadAsset(obj);
        }
    }
}

File: LuaScript/Library/Loader/Res/Assist/ResourcesLoader.lua
---
--- ResourcesLoader       
--- Author : hukiry     
--- DateTime 2024/8/8 14:26   
---

---@class ResourcesLoader
local ResourcesLoader = Class()
---初始化字段
function ResourcesLoader:ctor()
    self.OnLoadedCallback = nil
end

---@param OnLoadedCallback function<UnityEngine.Object>
function ResourcesLoader:LoadAsyncMp4(assetName, OnLoadedCallback)
    self.OnLoadedCallback = OnLoadedCallback
    StartCoroutine(Handle(self, self.LoadAssetAsync, assetName))
end

function ResourcesLoader:Unload(obj)
    UnityEngine.Resources.UnloadAsset(obj)
end

---@private
function ResourcesLoader:LoadAssetAsync(assetName)
    local rr = UnityEngine.Resources.LoadAsync(assetName);
    while (not rr.isDone) do
        WaitForEndOfFrame()
    end
    self.OnLoadedCallback(assetName, rr.asset)
    Yield(0)
end

return ResourcesLoader